gtsam/hybrid/HybridGaussianFactorGraph.cpp

```diff
@@ -130,49 +128,54 @@ GaussianFactorGraphTree HybridGaussianFactorGraph::assembleGraphTree() const {
 }
 
 /* ************************************************************************ */
-static std::pair<HybridConditional::shared_ptr, HybridFactor::shared_ptr>
+static std::pair<HybridConditional::shared_ptr, boost::shared_ptr<Factor>>
 continuousElimination(const HybridGaussianFactorGraph &factors,
                       const Ordering &frontalKeys) {
   GaussianFactorGraph gfg;
-  for (auto &fp : factors) {
-    if (auto ptr = boost::dynamic_pointer_cast<HybridGaussianFactor>(fp)) {
-      gfg.push_back(ptr->inner());
-    } else if (auto ptr = boost::static_pointer_cast<HybridConditional>(fp)) {
-      gfg.push_back(
-          boost::static_pointer_cast<GaussianConditional>(ptr->inner()));
+  for (auto &f : factors) {
+    if (auto gf = dynamic_pointer_cast<GaussianFactor>(f)) {
```

Comment:
I think I see why we're doing `FactoGraph<Factor>` and using `Factor` everywhere instead of `HybridFactor`.

We can make our lives much easier by having the hierarchy:

```
Factor
- SymbolicFactor
- HybridFactor
  - GaussianFactor
  - DiscreteFactor 
```

I mentioned this in the meeting we had. `SymbolicFactor` is not handled by the machinery here so it would have to be a sibling of `HybridFactor` until we provide that capability. Else merging HybridFactor and Factor is also an option.

Reply:
Yeah, I opted for simplicity and with an eye towards GTSAM 5.

Reply:
`SymbolicFactor` is really *in my opinion* a way to reuse the elimination templates for symbolic factorization. Thus a `SymbolicFactor` in some sense is not a `Factor`...

Reply:
I don’t disagree. In hindsight, maybe a better option is to do the symbolic elimination just with the keys in the factors. But, not having to worry about making “working” numerical factor graphs to test out elimination methods is very useful. One option is to make the Factor base class non-abstract so it can “take over” for symbolic.